tests/unit/obuf-limits.tcl

```diff
@@ -70,4 +70,32 @@ start_server {tags {"obuf-limits"}} {
         assert {$omem >= 100000 && $time_elapsed < 6}
         $rd1 close
     }
+
+    test {Won't write replies if client output buffer hard limit is enforced} {
+        r config set client-output-buffer-limit {normal 100000 0 0}
+
+        # Total size of all items must be more than 2 * 100000
+        for {set i 0} {$i < 100000} {incr i} {
+            r lpush mylist $i $i
+        }
+        set orig_mem [s used_memory]
+
+        # Get all items
+        set rd [redis_deferring_client]
+        $rd lrange mylist 0 -1
+
+        # Redis will not use 2 * 100000 memory to reply because we won't
+        # write any things if reached client output buffer hard limit.
+        set current_mem [s used_memory]
+        assert {$current_mem < 2 * 100000 + $orig_mem}
```

Comment:
I think you meant to test something else.
This test can test one of 3 things:
1. test that the used memory didn't grow to hold the entire LRANGE reply (including all elements). i.e. 200k in your test?
2. test that the used memory didn't grow beyond the output buffer limit. i.e. 100k in your test?
3. test that the used memory didn't grow at all (using some 10k threshold), because the client got disconnected and the 100k it cached got released.

i think you need to test 3, but the code seems to test 1.

Reply:
Yes, I want to test 3. My way to get used memory is wrong, but it seems we can't get used memory by info command, because the client memory will be released when redis executes a next command. I want to use debug command following lrange comman by pipeline, but redis doesn't read the client query buffer if it client is set 'CLIENT_CLOSE_ASYNC'.

Reply:
you can't put any other command in a pipeline after the LRANGE (it won't be executed), and if you use another client, you can't control the order of execution.
I don't understand what's wrong in just using INFO after we know the client was supposed to be disconnected (100ms after sending the LRANGE, and before reading from that socket)?
i.e. we are expecting to see that all the memory it cached in the output buffer before being flagged to be closed was released (even if it didn't attempt to read it yet)

Reply:
Oh, I originally want to test redis won't keep increasing memory even if there still are some reply bulks that need to add reply buffer when reached limit. But i think it is hard to test.
For the case you said, i will cover it.